File: tests/integration/csharp/HotReloadTest.cs
// tests/integration/csharp/HotReloadTest.cs
// Unit tests for ReloadPhase and RuntimeConfig types.
//
// Run with: dotnet test

using System;
using Polyplug.Host;

namespace Polyplug.Tests;

/// <summary>
/// Unit tests for ReloadPhaseType enum constants.
/// </summary>
public static class ReloadPhaseTypeTests
{
    public static void TypePreparingIs0()
    {
        if ((uint)ReloadPhaseType.Preparing != 0U)
        {
            throw new Exception("TYPE_PREPARING should be 0");
        }
    }

    public static void TypeReloadedIs1()
    {
        if ((uint)ReloadPhaseType.Reloaded != 1U)
        {
            throw new Exception("TYPE_RELOADED should be 1");
        }
    }

    public static void TypeFailedIs2()
    {
        if ((uint)ReloadPhaseType.Failed != 2U)
        {
            throw new Exception("TYPE_FAILED should be 2");
        }
    }
}

/// <summary>
/// Unit tests for ReloadPhase class.
/// </summary>
public static class ReloadPhaseTests
{
    public static void ConstructorSetsAllProperties()
    {
        var phase = new ReloadPhase(
            ReloadPhaseType.Preparing,
            12345UL,
            "TestBundle",
            2u,
            "Test reason"
        );

        if (phase.Type != ReloadPhaseType.Preparing)
        {
            throw new Exception("phase.Type should be Preparing");
        }
        if (phase.BundleId != 12345UL)
        {
            throw new Exception("phase.BundleId should be 12345");
        }
        if (phase.BundleName != "TestBundle")
        {
            throw new Exception("phase.BundleName should be TestBundle");
        }
        if (phase.RetryCount != 2u)
        {
            throw new Exception("phase.RetryCount should be 2");
        }
        if (phase.Reason != "Test reason")
        {
            throw new Exception("phase.Reason should be Test reason");
        }
    }

    public static void ConstructorUsesDefaultValues()
    {
        var phase = new ReloadPhase(
            ReloadPhaseType.Reloaded,
            999UL,
            "MyBundle"
        );

        if (phase.Type != ReloadPhaseType.Reloaded)
        {
            throw new Exception("phase.Type should be Reloaded");
        }
        if (phase.BundleId != 999UL)
        {
            throw new Exception("phase.BundleId should be 999");
        }
        if (phase.BundleName != "MyBundle")
        {
            throw new Exception("phase.BundleName should be MyBundle");
        }
        if (phase.RetryCount != 0u)
        {
            throw new Exception("phase.RetryCount should default to 0");
        }
        if (phase.Reason != "")
        {
            throw new Exception("phase.Reason should default to empty string");
        }
    }

    public static void ConstructorHandlesNullBundleName()
    {
        var phase = new ReloadPhase(
            ReloadPhaseType.Failed,
            1UL,
            null!,
            0u,
            "Error"
        );

        if (phase.BundleName != "")
        {
            throw new Exception("null bundleName should default to empty string");
        }
    }

    public static void ConstructorHandlesNullReason()
    {
        var phase = new ReloadPhase(
            ReloadPhaseType.Failed,
            1UL,
            "Bundle",
            0u,
            null!
        );

        if (phase.Reason != "")
        {
            throw new Exception("null reason should default to empty string");
        }
    }

    public static void IsPreparingReturnsTrueForPreparing()
    {
        var phase = new ReloadPhase(ReloadPhaseType.Preparing, 1UL, "Bundle");
        if (!phase.IsPreparing())
        {
            throw new Exception("IsPreparing should return true for Preparing");
        }
    }

    public static void IsPreparingReturnsFalseForReloaded()
    {
        var phase = new ReloadPhase(ReloadPhaseType.Reloaded, 1UL, "Bundle");
        if (phase.IsPreparing())
        {
            throw new Exception("IsPreparing should return false for Reloaded");
        }
    }

    public static void IsReloadedReturnsTrueForReloaded()
    {
        var phase = new ReloadPhase(ReloadPhaseType.Reloaded, 1UL, "Bundle");
        if (!phase.IsReloaded())
        {
            throw new Exception("IsReloaded should return true for Reloaded");
        }
    }

    public static void IsReloadedReturnsFalseForPreparing()
    {
        var phase = new ReloadPhase(ReloadPhaseType.Preparing, 1UL, "Bundle");
        if (phase.IsReloaded())
        {
            throw new Exception("IsReloaded should return false for Preparing");
        }
    }

    public static void IsFailedReturnsTrueForFailed()
    {
        var phase = new ReloadPhase(ReloadPhaseType.Failed, 1UL, "Bundle");
        if (!phase.IsFailed())
        {
            throw new Exception("IsFailed should return true for Failed");
        }
    }

    public static void IsFailedReturnsFalseForPreparing()
    {
        var phase = new ReloadPhase(ReloadPhaseType.Preparing, 1UL, "Bundle");
        if (phase.IsFailed())
        {
            throw new Exception("IsFailed should return false for Preparing");
        }
    }

    public static void ToStringIncludesAllRelevantFields()
    {
        var phase = new ReloadPhase(
            ReloadPhaseType.Preparing,
            42UL,
            "TestBundle",
            3u,
            "Ignored"
        );
        var result = phase.ToString();

        if (!result.Contains("Preparing"))
        {
            throw new Exception("ToString should include Preparing");
        }
        if (!result.Contains("42"))
        {
            throw new Exception("ToString should include BundleId");
        }
        if (!result.Contains("TestBundle"))
        {
            throw new Exception("ToString should include BundleName");
        }
        if (!result.Contains("3"))
        {
            throw new Exception("ToString should include RetryCount");
        }
    }
}

/// <summary>
/// Test runner entry point.
/// </summary>
public static class TestRunner
{
    private static int _passed = 0;
    private static int _failed = 0;

    private static void RunTest(Action test, string name)
    {
        try
        {
            test();
            _passed++;
            Console.WriteLine($"  PASS: {name}");
        }
        catch (Exception e)
        {
            _failed++;
            Console.WriteLine($"  FAIL: {name}");
            Console.WriteLine($"        {e.Message}");
        }
    }

    public static int Main()
    {
        Console.WriteLine("=== ReloadPhase Type Constants ===");
        RunTest(ReloadPhaseTypeTests.TypePreparingIs0, "TypePreparingIs0");
        RunTest(ReloadPhaseTypeTests.TypeReloadedIs1, "TypeReloadedIs1");
        RunTest(ReloadPhaseTypeTests.TypeFailedIs2, "TypeFailedIs2");

        Console.WriteLine("\n=== ReloadPhase Constructor ===");
        RunTest(ReloadPhaseTests.ConstructorSetsAllProperties, "ConstructorSetsAllProperties");
        RunTest(ReloadPhaseTests.ConstructorUsesDefaultValues, "ConstructorUsesDefaultValues");
        RunTest(ReloadPhaseTests.ConstructorHandlesNullBundleName, "ConstructorHandlesNullBundleName");
        RunTest(ReloadPhaseTests.ConstructorHandlesNullReason, "ConstructorHandlesNullReason");

        Console.WriteLine("\n=== ReloadPhase Helper Methods ===");
        RunTest(ReloadPhaseTests.IsPreparingReturnsTrueForPreparing, "IsPreparingReturnsTrueForPreparing");
        RunTest(ReloadPhaseTests.IsPreparingReturnsFalseForReloaded, "IsPreparingReturnsFalseForReloaded");
        RunTest(ReloadPhaseTests.IsReloadedReturnsTrueForReloaded, "IsReloadedReturnsTrueForReloaded");
        RunTest(ReloadPhaseTests.IsReloadedReturnsFalseForPreparing, "IsReloadedReturnsFalseForPreparing");
        RunTest(ReloadPhaseTests.IsFailedReturnsTrueForFailed, "IsFailedReturnsTrueForFailed");
        RunTest(ReloadPhaseTests.IsFailedReturnsFalseForPreparing, "IsFailedReturnsFalseForPreparing");

        Console.WriteLine("\n=== ReloadPhase ToString ===");
        RunTest(ReloadPhaseTests.ToStringIncludesAllRelevantFields, "ToStringIncludesAllRelevantFields");

        Console.WriteLine("\n========================================");
        Console.WriteLine($"Tests passed: {_passed}");
        Console.WriteLine($"Tests failed: {_failed}");
        Console.WriteLine("========================================");

        return _failed > 0 ? 1 : 0;
    }
}

File: tests/integration/cpp/hot_reload_test.cpp
// tests/integration/cpp/hot_reload_test.cpp
// Unit tests for the ReloadPhase type and its phase-type constants.
//
// Compile with: g++ -std=c++17 -I../../../sdks/cpp/abi hot_reload_test.cpp -o hot_reload_test
// Run: ./hot_reload_test

#include <cassert>
#include <cstdint>
#include <cstring>
#include <iostream>

#include "polyplug/abi.hpp"

static int tests_passed = 0;
static int tests_failed = 0;

#define ASSERT_EQUALS(expected, actual, message) \
    do { \
        if ((expected) == (actual)) { \
            tests_passed++; \
        } else { \
            tests_failed++; \
            std::cerr << "FAIL: " << (message) << std::endl; \
            std::cerr << "  Expected: " << (expected) << std::endl; \
            std::cerr << "  Actual: " << (actual) << std::endl; \
        } \
    } while (0)

// ============================================================================
// ReloadPhase Type Constants Tests
// ============================================================================

void test_reload_phase_type_constants() {
    std::cout << "=== ReloadPhase Type Constants ===" << std::endl;

    ASSERT_EQUALS(0U, static_cast<uint32_t>(ReloadPhaseType::Preparing), "TYPE_PREPARING should be 0");
    ASSERT_EQUALS(1U, static_cast<uint32_t>(ReloadPhaseType::Reloaded), "TYPE_RELOADED should be 1");
    ASSERT_EQUALS(2U, static_cast<uint32_t>(ReloadPhaseType::Failed), "TYPE_FAILED should be 2");
}

// ============================================================================
// ReloadPhase Struct Tests
// ============================================================================

void test_reload_phase_struct() {
    std::cout << "\n=== ReloadPhase Struct ===" << std::endl;

    // Test struct with all fields set
    ReloadPhase phase;
    phase.phase_type = ReloadPhaseType::Preparing;
    phase.bundle_id = 12345ULL;
    phase.bundle_name = StringView{reinterpret_cast<const uint8_t*>("TestBundle"), 10};
    phase.reason = StringView{reinterpret_cast<const uint8_t*>("Test reason"), 11};

    ASSERT_EQUALS(static_cast<uint32_t>(ReloadPhaseType::Preparing), static_cast<uint32_t>(phase.phase_type), "phase.phase_type should be Preparing");
    ASSERT_EQUALS(12345ULL, phase.bundle_id, "phase.bundle_id should be 12345");
    ASSERT_EQUALS(10U, phase.bundle_name.len, "phase.bundle_name.len should be 10");
    ASSERT_EQUALS(11U, phase.reason.len, "phase.reason.len should be 11");

    // Test struct with default-like values
    ReloadPhase phase_defaults;
    phase_defaults.phase_type = ReloadPhaseType::Reloaded;
    phase_defaults.bundle_id = 999ULL;
    phase_defaults.bundle_name = StringView{reinterpret_cast<const uint8_t*>("MyBundle"), 8};
    phase_defaults.reason = StringView{nullptr, 0};

    ASSERT_EQUALS(static_cast<uint32_t>(ReloadPhaseType::Reloaded), static_cast<uint32_t>(phase_defaults.phase_type), "phase_defaults.phase_type should be Reloaded");
    ASSERT_EQUALS(999ULL, phase_defaults.bundle_id, "phase_defaults.bundle_id should be 999");
    ASSERT_EQUALS(0U, phase_defaults.reason.len, "phase_defaults.reason.len should be 0");

    // Test Failed phase with reason
    ReloadPhase failed_phase;
    failed_phase.phase_type = ReloadPhaseType::Failed;
    failed_phase.bundle_id = 1ULL;
    failed_phase.bundle_name = StringView{reinterpret_cast<const uint8_t*>("FailedBundle"), 12};
    failed_phase.reason = StringView{reinterpret_cast<const uint8_t*>("Error occurred"), 14};

    ASSERT_EQUALS(static_cast<uint32_t>(ReloadPhaseType::Failed), static_cast<uint32_t>(failed_phase.phase_type), "failed_phase.phase_type should be Failed");
    ASSERT_EQUALS(14U, failed_phase.reason.len, "failed_phase.reason.len should be 14");
}

// ============================================================================
// Results
// ============================================================================

int main() {
    test_reload_phase_type_constants();
    test_reload_phase_struct();

    std::cout << "\n========================================" << std::endl;
    std::cout << "Tests passed: " << tests_passed << std::endl;
    std::cout << "Tests failed: " << tests_failed << std::endl;
    std::cout << "========================================" << std::endl;

    return tests_failed > 0 ? 1 : 0;
}
